File: 1636-sort-array-by-increasing-frequency/1636-sort-array-by-increasing-frequency.ts
function frequencySort(nums: number[]): number[] {
  const freq = new Map<number, number>();
  nums.forEach((e) => {
    if (freq.has(e)) {
      freq.set(e, freq.get(e)! + 1);
      return;
    }
    freq.set(e, 1);
  });

  const arr = [...freq.entries()].sort((p1, p2) => {
    if (p1[1] != p2[1]) {
      return p1[1] - p2[1];
    }
    return p2[0] - p1[0];
  });

  const result: number[] = [];

  arr.forEach((v) => {
    for (let i = 0; i < v[1]; i++) {
      result.push(v[0]);
    }
  });

  return result;
}

File: 1636-sort-array-by-increasing-frequency/1636-sort-array-by-increasing-frequency.kt
class Solution {
    fun frequencySort(nums: IntArray): IntArray {
        val freq = mutableMapOf<Int, Int>()
        nums.forEach {
            freq[it] = freq.getOrDefault(it, 0) + 1
        }

        val sortedPairs = freq.toList().sortedWith(Comparator compare@{ o1, o2 ->
            if (o1.second == o2.second) return@compare o2.first - o1.first
            return@compare o1.second - o2.second
        })

        val result = IntArray(nums.size) { 0 }

        var i = 0
        var j = 0
        while (i < sortedPairs.size) {
            val key = sortedPairs[i].first
            val f = sortedPairs[i].second

            for (k in 0..<f) {
                result[j] = key
                j++
            }

            i++
        }

        return result
    }

}